src/main/java/org/wildfly/security/ssl/X509CRLTrustManager.java

```diff
@@ -0,0 +1,149 @@
+     * @param maxCertPath sets the value of the maximum number of non-self-issued intermediate
+     *                    certificates that may exist in a certification path. The value must be equal or greater than 1.
+     * @param acceptedIssuers an array of certificate authority certificates which are trusted for authenticating peers (may be {@code null}).
+     */
+    public X509CRLTrustManager(KeyStore trustStore, InputStream crlFile, int maxCertPath, X509Certificate[] acceptedIssuers) {
+        checkNotNullParam("trustStore", trustStore);
+        checkMinimumParameter("maxCertPath", 1, maxCertPath);
+        try {
+            TrustManagerFactory tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+
+            tmf.init(trustStore);
+
+            PKIXBuilderParameters params = new PKIXBuilderParameters(trustStore, new X509CertSelector());
+
+            if (crlFile != null) {
+                CertStoreParameters csp = new CollectionCertStoreParameters(getCRLs(crlFile));
+                CertStore store = CertStore.getInstance("Collection", csp);
+                params.addCertStore(store);
+            }
+
+            params.setRevocationEnabled(true);
+            params.setMaxPathLength(maxCertPath);
+
+            tmf.init(new CertPathTrustManagerParameters(params));
+
+            this.trustManagers = tmf.getTrustManagers();
+        } catch (GeneralSecurityException e) {
+            throw ElytronMessages.log.sslErrorCreatingTrustManager(getClass().getName(), e);
+        }
+        this.acceptedIssuers = acceptedIssuers;
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param trustStore a {@link KeyStore} with the trusted certificates (must not be {@code null})
+     * @param crlFile the input stream pointing to a certificate revocation list (may be {@code null})
+     */
+    public X509CRLTrustManager(KeyStore trustStore, InputStream crlFile) {
+        this(trustStore, crlFile, DEFAULT_MAX_CERT_PATH_LENGTH, null);
+    }
+
+    /**
+     * <p>Creates a new instance. When using this constructor, the instance is going to obtain CRLs from the distribution points
+     * within the certificates being validated.
+     *
+     * <p>Make sure you have system property <code>com.sun.security.enableCRLDP</code> set.
+     *
+     * @param trustStore a {@link KeyStore} with the trusted certificates (must not be {@code null})
+     */
+    public X509CRLTrustManager(KeyStore trustStore) {
+        this(trustStore, null, DEFAULT_MAX_CERT_PATH_LENGTH, null);
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        for (TrustManager trustManager : trustManagers) {
+            if (trustManager instanceof X509TrustManager) {
+                X509TrustManager.class.cast(trustManager).checkClientTrusted(chain, authType);
```

Comment:
TBH in practice these arrays probably always just contain a single X509TrustManager, however if we know that for every single call (and one of these methods will be called for every connection) we only want the X509TrustManagers it may be an idea to filter the array in advance.